trie-db/src/node.rs

```diff
@@ -41,88 +42,112 @@ pub enum Node<'a> {
 	/// Branch node with support for a nibble (when extension nodes are not used).
 	NibbledBranch(NibbleSlice<'a>, [Option<&'a [u8]>; nibble_ops::NIBBLE_LENGTH], Option<&'a [u8]>),
 }
-/// A Sparse (non mutable) owned vector struct to hold branch keys and value
-#[cfg_attr(feature = "std", derive(Debug))]
+
+/// A `NodePlan` is a blueprint for decoding a node from a byte slice. The `NodePlan` is created
+/// by parsing an encoded node and can be reused multiple times. This is useful as a `Node` borrows
+/// from a byte slice and this struct does not.
+///
+/// The enum values mirror those of `Node` except that instead of byte slices, this struct stores
+/// ranges that can be used to index into a large byte slice.
 #[derive(Eq, PartialEq, Clone)]
-pub struct Branch {
-	data: Vec<u8>,
-	ubounds: [usize; 18],
-	has_value: bool,
+#[cfg_attr(feature = "std", derive(Debug))]
+pub enum NodePlan {
+	/// Null trie node; could be an empty root or an empty branch entry.
+	Empty,
+	/// Leaf node; has key slice and value. Value may not be empty.
+	Leaf {
+		partial: Range<usize>,
+		partial_padding: usize,
+		value: Range<usize>,
+	},
+	/// Extension node; has key slice and node data. Data may not be null.
+	Extension {
+		partial: Range<usize>,
+		partial_padding: usize,
+		child_data: Range<usize>,
+	},
+	/// Branch node; has slice of child nodes (each possibly null)
+	/// and an optional immediate node data.
+	Branch {
+		value: Option<Range<usize>>,
+		children: [Option<Range<usize>>; nibble_ops::NIBBLE_LENGTH],
```

Comment:
we are using a bit useless memory here: can be twice as small approximately, just putting index into slice, this seems to favor code readability over optimization that does not really make a difference. I am not sure where to stand here.

Reply:
Well, in this case all the memory is on the stack whereas a more memory-efficient alternative would require a dynamically sized Vec, requiring a heap allocation. This also automatically ensures that none of the indices are out of range. On the other hand, arrays are kind of a pain to work with since they are not `IntoIterator`. IMO, it's not worth changing though.

Reply:
I was just suggesting to store only one index per children. But it relies on assumption that each children are separated by a constant number of byte (1 in substrate).
But such assumption make codec trait less practical, it is not a good idea.